src/parser.js

```diff
@@ -34,99 +46,171 @@ class Parser {
+  *walk(tree, level) {
     if (!tree) {
       return
     }
 
     for (let i = 0; i < tree.length; i++) {
       const node = tree[i]
 
-      this.pug += `\n${this.parseNode(node, level)}`
+      const newline = this.parseNode(node, level)
+      if (newline) {
+        this.pug += `\n${newline}`
+      }
 
       if (
         node.childNodes &&
         node.childNodes.length > 0 &&
-        !this.isUniqueTextNode(node)
+        !hasSingleTextNodeChild(node)
       ) {
-        yield * this.walk(node.childNodes, level + 1)
+        yield* this.walk(node.childNodes, level + 1)
       }
     }
   }
 
-  parseComment (node, indent) {
-    const comment = node.data.split('\n')
-
-    // Differentiate single line to multi-line comments
-    if (comment.length > 1) {
-      const multiLine = comment.map(line => `${indent}  ${line}`).join('\n')
-      return `${indent}//${multiLine}`
-    } else {
-      return `${indent}//${comment}`
-    }
-  }
-
-  parseNode (node, level) {
-    const indent = this.indent.repeat(level)
+  /*
+   * Returns a Pug node name with all attributes set in parentheses.
+   */
+  static getNodeWithAttributes(node) {
+    const { tagName, attrs } = node
+    const attributes = []
+    let pugNode = tagName
 
-    if (isDocumentTypeNode(node)) {
-      return `${indent}doctype html`
+    if (!attrs) {
+      return pugNode
     }
 
-    if (isTextNode(node)) {
-      return `${indent}| ${node.value}`
-    } else if (isCommentNode(node)) {
-      return this.parseComment(node, indent)
-    } else if (isElementNode(node)) {
-      let line = `${indent}${this.setAttributes(node)}`
+    // Add CSS selectors to pug node and append any element attributes to it
+    for (const attr of attrs) {
+      const { name, value } = attr
 
-      if (this.isUniqueTextNode(node)) {
-        line += ` ${node.childNodes[0].value}`
+      // Remove div tag if a selector is present (shorthand)
+      // e.g. div#form() -> #form()
+      const hasSelector = name === 'id' || name === 'class'
+      if (tagName === DIV_NODE && hasSelector) {
+        pugNode = pugNode.replace(DIV_NODE, '')
       }
 
-      return line
-    } else {
-      return node
-    }
-  }
-
-  setAttributes (node) {
-    const { tagName, attrs: attributes } = node
-    const attributeList = []
-    let pugNode = tagName
-
-    attributes.forEach(({ name, value }) => {
-      let shorten = false
-
       switch (name) {
         case 'id':
-          shorten = true
           pugNode += `#${value}`
           break
         case 'class':
-          shorten = true
           pugNode += `.${value.split(' ').join('.')}`
           break
-        default:
-          attributeList.push(`${name}='${value}'`)
+        default: {
+          // Add escaped single quotes (\') to attribute values
+          const val = value.replace(/'/g, "\\'")
+          if (!val) {
+            attributes.push(name)
+          } else {
+            attributes.push(`${name}='${val}'`)
```

Comment:
you need to push either way, it's the value that is conditional
```
const attribute = !val ?  name : `${name}='${val}'`;

attributes.push(attribute);
```

Reply:
Good shout. Ternary makes much more sense here.